packages/shell-evaluator/src/shell-evaluator.ts

```diff
@@ -112,12 +120,19 @@ class ShellEvaluator {
         }
         return;
       default:
+        this.saveState();
         const rewrittenInput = this.asyncWriter.compile(input);
         this.bus.emit(
           'mongosh:rewrittenAsyncInput',
           { original: input.trim(), rewritten: rewrittenInput.trim() }
         );
-        return originalEval(rewrittenInput, context, filename);
+        try {
+          return await originalEval(rewrittenInput, context, filename);
+        } catch (err) {
+          // This is for browser/Compass
+          this.revertState();
+          throw err;
```

Comment:
To debug it try to throw a different error here (`new Error('revertState was called')`), should appear in the shell output

Reply:
Ty!